pkg/apis/picchu/v1alpha1/releasemanager_types.go

```diff
@@ -38,43 +39,74 @@ type ReleaseManagerSpec struct {
 }
 
-type ReleaseManagerReleaseStatus struct {
-	Tag            string       `json:"tag"`
-	LastUpdate     *metav1.Time `json:"lastUpdated"`
-	GitTimestamp   *metav1.Time `json:"gitTimestamp"`
-	CurrentPercent uint32       `json:"currentPercent"`
-	PeakPercent    uint32       `json:"peakPercent"`
-	Released       bool         `json:"released"`
-	Expired        bool         `json:"expired"`
+type ReleaseManagerRevisionStatus struct {
+	Tag            string                                 `json:"tag"`
+	LastUpdated    *metav1.Time                           `json:"lastUpdated"`
+	CurrentPercent uint32                                 `json:"currentPercent"`
+	PeakPercent    uint32                                 `json:"peakPercent"`
+	Released       bool                                   `json:"released"`
+	Expired        bool                                   `json:"expired"`
+	EverReleased   bool                                   `json:"everReleased"`
+	Deployed       bool                                   `json:"deployed"`
+	EverDeployed   bool                                   `json:"everDeployed"`
+	Retired        bool                                   `json:"retired"`
+	Resources      []ReleaseManagerRevisionResourceStatus `json:"resources"`
+	Scale          ReleaseManagerRevisionScaleStatus      `json:"scale"`
 }
 
-func (r *ReleaseManager) ReleaseStatus(incarnation Incarnation) *ReleaseManagerReleaseStatus {
-	for _, s := range r.Status.Releases {
-		if s.Tag == incarnation.Spec.App.Tag {
+type ReleaseManagerRevisionResourceStatus struct {
+	ApiVersion string                `json:"apiVersion"`
+	Kind       string                `json:"kind"`
+	Metadata   *types.NamespacedName `json:"metadata,omitempty"`
+	Status     string                `json:"status"`
+}
+
+func (r *ReleaseManagerRevisionStatus) CreateOrUpdateResourceStatus(rs ReleaseManagerRevisionResourceStatus) {
+	for i, s := range r.Resources {
+		if s.Metadata.Name == rs.Metadata.Name &&
+			s.Metadata.Namespace == rs.Metadata.Namespace &&
+			s.ApiVersion == rs.ApiVersion &&
+			s.Kind == rs.Kind {
+			r.Resources[i].Status = rs.Status
+			return
+		}
+
+	}
+	r.Resources = append(r.Resources, rs)
+	return
+}
+
+type ReleaseManagerRevisionScaleStatus struct {
+	Current int32
+	Desired int32
+	Peak    int32
+}
+
+func (r *ReleaseManager) RevisionStatus(tag string) *ReleaseManagerRevisionStatus {
+	for _, s := range r.Status.Revisions {
+		if s.Tag == tag {
 			return &s
 		}
 	}
 	now := metav1.Now()
-	gitTimestamp := metav1.NewTime(incarnation.GitTimestamp())
-	s := ReleaseManagerReleaseStatus{
-		Tag:            incarnation.Spec.App.Tag,
-		LastUpdate:     &now,
+	s := ReleaseManagerRevisionStatus{
+		Tag:            tag,
 		Expired:        false,
-		GitTimestamp:   &gitTimestamp,
+		LastUpdated:    &now,
 		CurrentPercent: 0,
 		PeakPercent:    0,
 	}
-	r.Status.Releases = append(r.Status.Releases, s)
+	r.Status.Revisions = append(r.Status.Revisions, s)
```

Comment:
should these be sorted?

Reply:
yeah, I've been thinking about it. another PR